File: examples/spaced/py/commands/daemon.py
"""commands/daemon.py — daemon list, status <id>, stop <id>, stop --all."""
from __future__ import annotations

import sys
from typing import Optional

import typer

sys.path.insert(0, __import__("os").path.join(__import__("os").path.dirname(__file__), "../../../../sdk/py"))
from hop_top_kit.bus import Bus, create_event  # noqa: E402
from data import DAEMONS, find_daemon  # noqa: E402

app = typer.Typer(help="Manage background controversy processes", add_completion=False)

# Module-level bus; set by spaced.py before command runs.
_bus: Bus | None = None

_NEW_DAEMON_NAME = "musk-response-to-this-cli"

_STATUS_ICONS = {
    "RUNNING": "●",
    "STOPPED": "○",
    "PAUSED": "◐",
}


@app.command("list")
def daemon_list() -> None:
    """List all known background controversy daemons."""
    typer.echo(f"\n  {'ID':<35}{'STATUS':<10}STARTED")
    typer.echo(f"  {'─' * 60}")
    for d in DAEMONS:
        icon = _STATUS_ICONS.get(d.status, "?")
        typer.echo(f"  {icon} {d.id:<33}{d.status:<10}{d.started}")
    typer.echo("")
    typer.echo(f"  {len(DAEMONS)} daemons running. 0 resolved. Trend: upward.")
    typer.echo("")


@app.command("status")
def daemon_status(
    daemon_id: str = typer.Argument(..., help="Daemon ID (see daemon list)"),
) -> None:
    """Display status and details for a specific daemon."""
    d = find_daemon(daemon_id)
    if d is None:
        typer.echo(f"  ✗ Daemon not found: {daemon_id}", err=True)
        raise typer.Exit(1)

    icon = _STATUS_ICONS.get(d.status, "?")
    refs_str = "\n".join(f"    → {r}" for r in d.refs)

    typer.echo(f"""
  {icon} DAEMON: {d.name}
  {"─" * 56}
  ID:       {d.id}
  Status:   {d.status}
  Started:  {d.started}

  Description:
    {d.description}

  References:
{refs_str}

  Note: {d.notes}
""")


@app.command("stop")
def daemon_stop(
    daemon_id: Optional[str] = typer.Argument(None, help="Daemon ID to stop (omit with --all)"),
    all: bool = typer.Option(False, "--all", help="Attempt to stop all daemons"),
    force: bool = typer.Option(False, "--force", help="Bypass safety guard"),
) -> None:
    """Attempt to stop a controversy daemon. Results may vary."""
    from hop_top_kit.safety import SafetyLevel, safety_guard  # noqa: PLC0415

    level = SafetyLevel.DANGEROUS if all else SafetyLevel.CAUTION
    safety_guard(level, force=force)

    if all:
        count = len(DAEMONS)
        typer.echo(f"""
  ✗ STOP FAILED: all daemons ({count}/{count})
  {"─" * 50}
  Stopped:                           0
  Still running:                     {count}
  New daemons spawned during attempt: 1
    → {_NEW_DAEMON_NAME}  [RUNNING since just now]

  Diagnosis:
    Controversies are self-sustaining processes. SIGTERM ignored.
    SIGKILL not permitted (First Amendment, probably).
    FAA, SEC, NLRB, and WSJ all notified. All still watching.

  Recommendation:
    Close this terminal. Touch grass. It will not help. But try.
""")
        return

    if not daemon_id:
        typer.echo("  ✗ Provide a daemon ID or use --all", err=True)
        raise typer.Exit(1)

    d = find_daemon(daemon_id)
    if d is None:
        typer.echo(f"  ✗ Daemon not found: {daemon_id}", err=True)
        raise typer.Exit(1)

    if _bus is not None:
        _bus.publish(create_event(
            "kit.spaced.daemon.stopped", "spaced", {"daemon": daemon_id},
        ))

    typer.echo(f"""
  ✗ STOP FAILED: {d.id}
  {"─" * 50}
  Daemon:   {d.name}
  Status:   {d.status} (unchanged)
  Signal:   SIGTERM (ignored)

  Stop attempt log:
    → sent SIGTERM... daemon tweeted in response
    → sent SIGKILL... daemon cited free speech
    → sent regulatory inquiry... added to queue (est. 18 months)
    → filed court motion... appeal pending

  The daemon will continue running.
  Refs: {d.refs[0] if d.refs else "see public record"}
""")

File: examples/spaced/ts/commands/daemon.ts
/**
 * commands/daemon.ts — daemon list / status <id> / stop <id> / stop --all
 */

import { Command } from 'commander';
import { Bus, createEvent } from '../../../../sdk/ts/src/bus';
import { DAEMONS, findDaemon } from '../data';

function statusIcon(status: string): string {
  switch (status) {
    case 'RUNNING':  return '●';
    case 'RESOLVED': return '○';
    default:         return '?';
  }
}

export function daemonCommand(bus?: Bus): Command {
  const daemon = new Command('daemon')
    .description('Manage background controversy processes');

  // --- daemon list ---
  daemon
    .command('list')
    .description('List all known controversy daemons')
    .action(function () {
      const COL = { id: 32, started: 14, status: 12 };

      const header =
        '  ' +
        'DAEMON ID'.padEnd(COL.id) +
        'STARTED'.padEnd(COL.started) +
        'STATUS';
      const sep = '  ' + '─'.repeat(COL.id + COL.started + COL.status + 4);

      console.log('');
      console.log(header);
      console.log(sep);

      for (const d of DAEMONS) {
        const row =
          '  ' +
          d.id.padEnd(COL.id) +
          d.started.padEnd(COL.started) +
          statusIcon(d.status) + ' ' + d.status;
        console.log(row);
      }

      const running  = DAEMONS.filter(d => d.status === 'RUNNING').length;
      const resolved = DAEMONS.filter(d => d.status === 'RESOLVED').length;

      console.log('');
      console.log(`  Running: ${running}   Resolved: ${resolved}   Total: ${DAEMONS.length}`);
      console.log('');
      console.log('  ● RUNNING  = Active, ongoing, legally significant');
      console.log('  ○ RESOLVED = Settled, concluded, archived (not forgotten)');
      console.log('');
    });

  // --- daemon status <id> ---
  daemon
    .command('status <id>')
    .description('Get detailed status of a daemon by ID')
    .action(function (id: string) {
      const d = findDaemon(id);

      if (!d) {
        process.stderr.write(`  error: unknown daemon: "${id}"\n`);
        process.stderr.write(`  Run 'spaced daemon list' to see available daemons.\n`);
        process.exit(1);
      }

      console.log('');
      console.log(`  DAEMON: ${d.id}`);
      console.log(`  ${'─'.repeat(50)}`);
      console.log(`  Title:   ${d.title}`);
      console.log(`  Status:  ${statusIcon(d.status)} ${d.status}`);
      console.log(`  Started: ${d.started}`);
      console.log('');
      console.log('  Description:');
      // Word-wrap at 70 chars
      const words = d.description.split(' ');
      let line = '  ';
      for (const word of words) {
        if (line.length + word.length + 1 > 72) {
          console.log(line);
          line = '  ' + word;
        } else {
          line += (line === '  ' ? '' : ' ') + word;
        }
      }
      if (line.trim()) console.log(line);

      console.log('');
      console.log('  References:');
      for (const ref of d.refs) {
        console.log(`    [${ref.outlet}] ${ref.headline}`);
        console.log(`           ${ref.url}`);
      }
      console.log('');
    });

  // --- daemon stop <id> / daemon stop --all ---
  daemon
    .command('stop [id]')
    .description('Attempt to stop a daemon (results may vary)')
    .option('--all', 'Stop all daemons simultaneously', false)
    .action(function (id: string | undefined) {
      const opts = this.opts();
      const stopAll = opts['all'] as boolean;

      if (stopAll) {
        const running = DAEMONS.filter(d => d.status === 'RUNNING');
        const newDaemon = {
          id: 'musk-response-to-this-cli',
          status: 'RUNNING',
        };

        console.log('');
        console.log(`  ✗ STOP FAILED: all daemons (${running.length}/${running.length})`);
        console.log(`  Stopped: 0`);
        console.log(`  Still running: ${running.length}`);
        console.log(`  New daemons spawned during stop attempt: 1`);
        console.log(`    → ${newDaemon.id}  [RUNNING since just now]`);
        console.log('');
        console.log('  The daemons are self-perpetuating. This is a known issue.');
        console.log('  Recommended action: document them. Stop filing for injunctions.');
        console.log('');
        return;
      }

      if (!id) {
        process.stderr.write('  error: provide a daemon ID or use --all\n');
        process.stderr.write('  Run \'spaced daemon list\' to see available daemons.\n');
        process.exit(1);
      }

      const d = findDaemon(id);

      if (!d) {
        process.stderr.write(`  error: unknown daemon: "${id}"\n`);
        process.stderr.write(`  Run 'spaced daemon list' to see available daemons.\n`);
        process.exit(1);
      }

      bus?.publish(createEvent(
        'kit.spaced.daemon.stopped', 'spaced',
        { daemon: id },
      ));

      if (d.status === 'RESOLVED') {
        console.log('');
        console.log(`  ○ DAEMON ALREADY RESOLVED: ${d.id}`);
        console.log(`  Resolved on: (see settlement terms, FOIA #TBD)`);
        console.log('  Nothing further to stop. The record remains.');
        console.log('');
        return;
      }

      console.log('');
      console.log(`  ✗ STOP FAILED: ${d.id}`);
      console.log(`  The daemon persists. Regulatory agencies, journalists,`);
      console.log(`  and approximately 400 million Twitter users keep it alive.`);
      console.log('');
      console.log(`  Suggested alternatives:`);
      console.log(`    → Settle out of court`);
      console.log(`    → Issue apology (optional, per legal)`);
      console.log(`    → Wait for news cycle to shift (avg: 3 days)`);
      console.log(`    → Acquire the outlet reporting it (expensive, impractical)`);
      console.log('');
    });

  return daemon;
}
